v2/spiffetls/tlsconfig/config_test.go

```diff
@@ -0,0 +1,687 @@
+			name:         "invalid bundle",
+			serverConfig: tlsconfig.TLSServerConfig(serverSVID),
+			clientConfig: tlsconfig.TLSClientConfig(bundle2, tlsconfig.AuthorizeMemberOf(td)),
+			err:          `x509svid: could not get X509 bundle: x509bundle: no X.509 bundle found for trust domain: "domain1.test"`,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		t.Run(testCase.name, func(t *testing.T) {
+			testConnection(t, testCase.serverConfig, testCase.clientConfig, testCase.err)
+		})
+	}
+}
+
+func TestMTLSHandshake(t *testing.T) {
+	// Create Bundle1
+	td := spiffeid.RequireTrustDomainFromString("domain1.test")
+	ca1 := test.NewCA(t)
+	bundle1 := ca1.Bundle(td)
+	require.NotNil(t, bundle1)
+
+	// Create Server SVID
+	svid1ID := td.NewID("server")
+	svid1Certs, key1 := ca1.CreateX509SVID(svid1ID.String())
+	require.NotNil(t, svid1Certs)
+	require.NotNil(t, key1)
+	serverSVID := &x509svid.SVID{
+		ID:           svid1ID,
+		Certificates: append(svid1Certs, ca1.Roots()...),
+		PrivateKey:   key1,
+	}
+
+	// Create Client SVID
+	//svid2ID := td.NewID("client")
+	//svid2Certs, key2 := ca1.CreateX509SVID(svid2ID.String())
+	//require.NotNil(t, svid1Certs)
+	//require.NotNil(t, key1)
+	//clientSVID := &x509svid.SVID{
+	//	ID:           svid2ID,
+	//	Certificates: svid2Certs,
+	//	PrivateKey:   key2,
+	//}
+
+	// Create Bundle2
+	td2 := spiffeid.RequireTrustDomainFromString("domain2.test")
+	ca2 := test.NewCA(t)
+	bundle2 := ca2.Bundle(td2)
+	require.NotNil(t, bundle2)
+
+	testCases := []struct {
+		name         string
+		serverConfig *tls.Config
+		clientConfig *tls.Config
+		err          string
+	}{
+		{
+			name:         "success",
+			serverConfig: tlsconfig.MTLSServerConfig(serverSVID, bundle1, tlsconfig.AuthorizeAny()),
+			clientConfig: tlsconfig.MTLSClientConfig(serverSVID, bundle1, tlsconfig.AuthorizeAny()),
+		},
+		//{
+		//	name:         "invalid trust domain",
+		//	serverConfig: tlsconfig.TLSServerConfig(serverSVID),
+		//	clientConfig: tlsconfig.TLSClientConfig(bundle1, tlsconfig.AuthorizeMemberOf(td2)),
+		//	err:          `unexpected trust domain "domain1.test"`,
+		//},
+		//{
+		//	name:         "invalid bundle",
+		//	serverConfig: tlsconfig.TLSServerConfig(serverSVID),
+		//	clientConfig: tlsconfig.TLSClientConfig(bundle2, tlsconfig.AuthorizeMemberOf(td)),
+		//	err:          `x509svid: could not get X509 bundle: x509bundle: no X.509 bundle found for trust domain: "domain1.test"`,
+		//},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		t.Run(testCase.name, func(t *testing.T) {
+			testConnection(t, testCase.serverConfig, testCase.clientConfig, testCase.err)
+		})
+	}
+}
+
+func testConnection(t testing.TB, serverConfig *tls.Config, clientConfig *tls.Config, expectedErr string) {
+	ln, err := tls.Listen("tcp", "127.0.0.1:0", serverConfig)
+	require.NoError(t, err)
+
+	go func() {
+		conn, err := ln.Accept()
+		require.NoError(t, err)
+		defer conn.Close()
+
+		r := bufio.NewReader(conn)
```

Comment:
nit: reading and writing strings seems to just add more code for no benefit. I'd suggest just reading and writing a single byte, since all we care is that the read/write succeed (to complete the TLS1.3 handshake).